feature(offchain): add Redis Cluster support

torives opened the pull request:
Closes #12

torives committed:
feat(offchain): add Redis Cluster support

gligneul commented on offchain/rollups-events/src/broker/mod.rs:
```diff
@@ -285,6 +354,7 @@ pub struct BrokerCLIConfig {
 #[derive(Debug, Clone)]
 pub struct BrokerConfig {
     pub redis_endpoint: RedactedUrl,
+    pub redis_cluster_endpoints: Vec<RedactedUrl>,
```
It is strange to have both options when only one is available. The type does not correspond to the semantics of the object. We should use an enum to represent these fields. For instance:
```
enum BrokerEndpoint {
    Single(RedactedUrl),
    Cluster(Vec<RedactedUrl>),
}
```
Then
```suggestion
    pub endpoint: BrokerEndpoint,
```

gligneul commented on offchain/rollups-events/src/broker/mod.rs:
```diff
@@ -273,6 +337,11 @@ pub struct BrokerCLIConfig {
 
+    /// Address list of Redis cluster nodes, defined by a single string
+    /// separated by empty spaces. If present, it supersedes `redis_endpoint`
+    #[arg(long, env, num_args = 1.., value_delimiter = ' ')]
```
Wouldn't it be better to use `,` as the delimiter?

torives replied:
Arguably, yes. I went with empty spaces to keep it similar to the `redis-cli` interface, but I have no objections to changing it.

gligneul commented on offchain/rollups-events/src/broker/mod.rs:
```diff
@@ -28,10 +32,55 @@ pub mod indexer;
 #[derive(Clone)]
 pub struct Broker {
-    connection: ConnectionManager,
+    connection: BrokerConnection,
```
Couldn't this be `Box<dyn ConnectionLike>` instead? If yes, you wouldn't need the BrokerConnection struct.

torives replied:
It's a tad bit more complicated than that. What we actually want is a type that has the [AsyncCommands](https://docs.rs/redis/0.23.0/redis/trait.AsyncCommands.html) and the `Clone` traits, as [ClusterClient](https://docs.rs/redis/0.23.0/redis/cluster/struct.ClusterClient.html) and [ConnectionManager](https://docs.rs/redis/0.23.0/redis/aio/struct.ConnectionManager.html) do.

I initially thought that a trait object was the answer, but then I found out that wouldn't work because `AsyncCommands` is only implemented for `Sized` types, and that breaks [object safety](https://doc.rust-lang.org/reference/items/traits.html#object-safety).

So, in the end, I used the same strategy we used for the [ConditionalSigner](https://github.com/cartesi/rollups/pull/124/files#diff-95a183c7c9178ec22d96a0c13f2e4c0e26f7fca31302d79de605c6604747d2c5R35) in #124.

gligneul commented on offchain/rollups-events/src/broker/mod.rs:
```diff
@@ -273,6 +337,11 @@ pub struct BrokerCLIConfig {
     #[arg(long, env, default_value = "redis://127.0.0.1:6379")]
     redis_endpoint: String,
 
+    /// Address list of Redis cluster nodes, defined by a single string
+    /// separated by empty spaces. If present, it supersedes `redis_endpoint`
```
We should document that the services will discover other nodes so that the user can define a single one here, and it will work.

gligneul commented on t:
```diff
@@ -0,0 +1,344 @@
+[37mdiff --git a/offchain/advance-runner/src/broker.rs b/offchain/advance-runner/src/broker.rs[m
```
Some garbage was committed

torives replied:
🫠

endersonmaia commented:
After starting a validator-node based on this branch connected to a AWS MemoryDB Redis cluster endpoint, I get the following errors:

- `rediss://clustercfg.rollups.gp1e3k.memorydb.us-east-1.amazonaws.com`

**indexer**

```
2023-07-18T17:49:16.897126Z  INFO indexer::indexer: connected to database; connecting to broker
2023-07-18T17:49:17.008070Z  INFO indexer::indexer: connected to broker; starting main loop
2023-07-18T17:49:17.008109Z  INFO indexer::indexer: waiting for next event self.state=IndexerState { inputs_last_id: "0", outputs_last_id: "0", inputs_stream: RollupsInputsStream { key: "chain-0:dapp-a057811d929
1cea1a6e7c0ff1eca12d6924b1de2:rollups-inputs" }, outputs_stream: RollupsOutputsStream { key: "chain-0:dapp-a057811d9291cea1a6e7c0ff1eca12d6924b1de2:rollups-outputs" } }
Error: BrokerError { source: ConnectionError { source: An error was signalled by the server: Keys in request don't hash to the same slot } }
```

**dispatcher**

```
2023-07-18T17:49:15.918153Z  INFO dispatcher::dispatcher: Setting up dispatcher with config: DispatcherConfig { sc_config: SCConfig { grpc_endpoint: "http://localhost:50051", default_confirmations: 1 }, tx_confi
g: TxManagerConfig { default_confirmations: 2, provider_http_endpoint: "https://eth-goerli.g.alchemy.com/v2/4eeMGl2u3ALzBEsohk3IuvNxSIPb0r0U", chain_id: 5, chain_is_legacy: false, database_path: "./default_tx_da
tabase", gas_oracle_api_key: "" }, broker_config: BrokerConfig { redis_endpoint: Cluster([rediss://clustercfg.rollups.gp1e3k.memorydb.us-east-1.amazonaws.com]), consume_timeout: 5000, backoff: ExponentialBackoff
 { current_interval: 500ms, initial_interval: 500ms, randomization_factor: 0.5, multiplier: 1.5, max_interval: 60s, start_time: Instant { tv_sec: 13176, tv_nsec: 860213054 }, max_elapsed_time: Some(120s), clock:
 SystemClock } }, auth_config: AWS { key_id: "2a313b66-752b-4787-bddf-bd59309fcdd1", region: UsEast1 }, dapp_deployment: DappDeployment { dapp_address: 0xa057811d9291cea1a6e7c0ff1eca12d6924b1de2, deploy_block_ha
sh: 0xd1e0419a922501c497feb4928d19f8acc4b123ff59218663bae0dd53b8cd5d54 }, rollups_deployment: RollupsDeployment { history_address: 0xa860738792a1e216cf57e7375d5a8637762538f5, authority_address: 0x5ba88473ac88ae1
55c1bd28cdd05f1a80e3bba18, input_box_address: 0x5a723220579c0dcb8c9253e6b4c62e572e379945 }, epoch_duration: 86400, priority: Normal }
Error: ConnectError { source: tonic::transport::Error(Transport, hyper::Error(Connect, ConnectError("tcp connect error", Os { code: 111, kind: ConnectionRefused, message: "Connection refused" }))) }
```

**advance-runner**

This doesn't look related to redis, but the container got restarted after the error, I don't know if should. In case it's really not related, we could open an issue to track that message.

```
2023-07-18T16:05:22.550177Z  INFO advance_runner: starting advance runner config=AdvanceRunnerConfig { server_manager_config: ServerManagerConfig { server_manager_endpoint: "http://localhost:5001", session_id: "
2023-07-18T16:05:22.731748Z  INFO advance_runner::runner: got latest snapshot snapshot=Snapshot { path: "/var/opt/cartesi/machine-snapshots/0_0", epoch: 0, processed_input_count: 0 }
2023-07-18T16:05:40.783821Z  INFO advance_runner::runner: starting runner main loop last_id="0"
2023-07-18T17:35:49.951759Z  INFO advance_runner::runner: consumed input event event=Event { id: "1689701749946-0", payload: RollupsInput { parent_id: "0", epoch_index: 0, inputs_sent_count: 1, data: AdvanceStat
2023-07-18T17:35:51.978968Z  INFO advance_runner::runner: waiting for the next input event last_id="1689701749946-0"
2023-07-18T17:35:51.979562Z  INFO advance_runner::runner: consumed input event event=Event { id: "1689701749953-0", payload: RollupsInput { parent_id: "1689701749946-0", epoch_index: 0, inputs_sent_count: 1, dat
2023-07-18T17:35:52.430664Z  INFO advance_runner::runner: produced epoch claim
2023-07-18T17:35:52.445187Z  INFO advance_runner::runner: waiting for the next input event last_id="1689701749953-0"
Error: RunnerFSSnapshotError { source: ParentIdMismatchError { expected: "1689701749953-0", got: "0" } }
```

torives committed:
feat(offchain): add hash tags to redis stream keys

When connecting to a Redis Cluster we must ensure that all keys in a request
are stored in the same hash slot. Defining the DApp's chain_id and
address as a hash tag for the streams solves that

gligneul commented on offchain/rollups-events/src/rollups_stream.rs:
```diff
@@ -73,7 +73,7 @@ impl From<DAppMetadataCLIConfig> for DAppMetadata {
 
 /// Declares a struct that implements the BrokerStream interface
-/// The generated key has the format `chain-<chain_id>:dapp-<dapp_address>:<key>`
+/// The generated key has the format `{chain-<chain_id>:dapp-<dapp_address>}:<key>`
```
Please add a comment explaining why we need the braces.

endersonmaia reviewed (approved):
LGTM :rocket: 

Testing the latest version I get no errors from `indexer`.

I made a comment about the CLI arguments at https://github.com/cartesi/rollups/issues/12#issuecomment-1640709004
